### PR TITLE
Fix race in RunCommandWithOutputForDuration

### LINKED ISSUE #18305: Flaky test: DockerCmdSuite.TestDockerCmdInDirWithTimeout
Randomly failing in almost every build (unit test)

see https://jenkins.dockerproject.org/job/Docker-PRs/20215/console

```
----------------------------------------------------------------------
FAIL: dockerCmd_utils_test.go:270: DockerCmdSuite.TestDockerCmdInDirWithTimeout

dockerCmd_utils_test.go:321:
    c.Assert(exitCode, check.Equals, cmd.expectedExitCode, check.Commentf("Expected exitCode %q for arguments %v, got %q", cmd.expectedExitCode, cmd.args, exitCode))
... obtained int = -1
... expected int = 0
... Expected exitCode '\x00' for arguments [a command that times out], got %!q(int=-1)

OOPS: 7 passed, 1 FAILED
--- FAIL: Test (0.05s)
FAIL
```

### PR DESCRIPTION
**Pull request by aaronlehmann**
This function was starting a goroutine that modifies one of its return
values. The intent is for the goroutine to only influence the return
value when it's causing the function to return, but it's racy and can
also modify the return value when the function is returning due to the
timeout. Fix the goroutine to not modify return values directly.

Also, give the channel a buffer so that the goroutine doesn't block
forever after a timeout.

Fixes #18305

### CODE REVIEW

File: pkg/integration/utils.go [hunk 1]
```diff
@@ -104,19 +104,25 @@ func RunCommandWithOutputForDuration(cmd *exec.Cmd, duration time.Duration) (out
...
 	if err != nil {
 		err = fmt.Errorf("Fail to start command %v : %v", cmd, err)
 	}
 
+	type exitInfo struct {
+		exitErr  error
+		exitCode int
+	}
+
+	done := make(chan exitInfo)
```
Comment (tonistiigi) [inline pkg/integration/utils.go]: nit: `buf=1` ?

Comment (aaronlehmann) [inline pkg/integration/utils.go]: Added a buffer. So now this fixes two concurrency bugs in this function.

**Comment by tonistiigi** [comment]
LGTM

**Comment by duglin** [comment]
LGTM thanks - just waiting on the tests

**Comment by LK4D4** [comment]
LGTM

**Comment by vdemeester** [comment]
@aaronlehmann :heart_eyes_cat:

**Pull request merged**